projects/plugins/jetpack/class.json-api-endpoints.php

```diff
@@ -2618,6 +2649,137 @@ public function get_amp_cache_origins( $siteurl ) {
+	/**
+	 * Handle the rest call.
+	 *
+	 * @param WP_REST_Request $request The request object.
+	 *
+	 * @return mixed|WP_Error
+	 */
+	public function rest_callback( WP_REST_Request $request ) {
+		$blog_id = Jetpack_Options::get_option( 'id' );
+
+		$this->api->initialize();
+		$this->api->endpoint = $this;
+
+		$response = call_user_func_array(
+			array( $this, 'callback' ),
+			array_values( array( $this->path, $blog_id ) + $request->get_url_params() )
+		);
+
+		$token_data = ( new Manager() )->verify_xml_rpc_signature();
+
+		if ( ! $token_data || empty( $token_data['token_key'] ) || ! array_key_exists( 'user_id', $token_data ) ) {
+			return new WP_Error( 'response_signature_error' );
+		}
+
+		$token = ( new Tokens() )->get_access_token( $token_data['user_id'], $token_data['token_key'] );
+		if ( is_wp_error( $token ) ) {
+			return $token;
+		}
+		if ( ! $token ) {
+			return new WP_Error( 'response_signature_error' );
+		}
+
+		$response = wp_json_encode( $response );
+		$nonce    = wp_generate_password( 10, false );
+		$hmac     = hash_hmac( 'sha1', $nonce . $response, $token->secret );
+
+		return array(
+			$response,
+			(string) $nonce,
+			(string) $hmac,
+		);
+	}
+
+	/**
+	 * The REST endpoint should only be available for requests signed with a valid blog or user token.
+	 * Declaring it "final" so individual endpoints couldn't remove this requirement.
+	 *
+	 * @return true|WP_Error
+	 */
+	final public function rest_permission_callback() {
+		$manager = new Manager( 'jetpack' );
+		if ( ! $manager->is_connected() ) {
+			return new WP_Error( 'site_not_connected' );
+		}
+
+		$user_id = Rest_Authentication::init()->wp_rest_authenticate( false );
+
+		$allow_blog_token = $this->allow_fallback_to_jetpack_blog_token || $this->allow_jetpack_site_auth;
+
+		if ( ( $allow_blog_token && Rest_Authentication::is_signed_with_blog_token() ) || ( $user_id && Rest_Authentication::is_signed_with_user_token() ) ) {
+			$custom_permission_result = $this->rest_permission_callback_custom();
+
+			// Successful custom permission check.
+			if ( $custom_permission_result === true ) {
+				if ( $user_id ) {
+					wp_set_current_user( $user_id );
+				}
+				return true;
+			}
+
+			// Custom permission check errored, returning the error.
+			if ( is_wp_error( $custom_permission_result ) ) {
+				return $custom_permission_result;
+			}
+
+			// Custom permission check failed, but didn't return a specific error. Proceed to returning the generic error.
+		}
+
+		$message = esc_html__(
+			'You do not have the correct user permissions to perform this action. Please contact your site admin if you think this is a mistake.',
+			'jetpack'
+		);
+		return new WP_Error( 'rest_api_invalid_permission', $message, array( 'status' => rest_authorization_required_code() ) );
+	}
+
+	/**
+	 * Redefine in individual endpoint classes to further customize the permission check.
+	 *
+	 * @return true|WP_Error
+	 */
+	public function rest_permission_callback_custom() {
```

Comment:
The idea is to make `rest_permission_callback()` method "final", so endpoints wouldn't be able to accidentally skip the authorization.
And `rest_permission_callback_custom()` is there to add some custom permission checks specifically for REST if they are ever needed.
Currently endpoints add them right into the `callback()` method, which I think is a bit messy, so this might be a way to improve them in the future.

Reply:
Yeah I see how you intended it to be used but I'm afraid in this case it might introduce more complexity to an already complex class.. For example, introducing duplicate capability checks or being misused on WPCOM..

Reply:
That's a valid point.
However, I believe if we have to choose between security checks being accidentally duplicated and accidentally removed, I'd rather have them duplicated 😅 

XML-RPC JSON API relies on centralized token checking that doesn't involve individual endpoints at all. They can add more checks, but cannot remove existing ones, so I'd prefer we kept it that way.

I improved the comments a bit to make the idea more clear: https://github.com/Automattic/jetpack/pull/39432/commits/79fdce830c39d958b012d4aec0a1d0d8b571d94e
But if somebody is confused, I'd rather have the main permission check `final` so they weren't able to accidentally remove it.